evap/rewards/tools.py

```diff
@@ -59,30 +59,45 @@ def reward_points_of_user(user):
     return SemesterActivation.objects.filter(semester=semester, is_active=True).exists()
 
-
-# Signal handlers
-
-@receiver(Course.course_evaluated)
-def grant_reward_points(sender, **kwargs):
+def grant_reward_points(user, semester):
     # grant reward points if all conditions are fulfilled
 
-    request = kwargs['request']
-    semester = kwargs['semester']
-    if not can_user_use_reward_points(request.user):
-        return
+    if not can_user_use_reward_points(user):
+        return False
     # has the semester been activated for reward points?
     if not is_semester_activated(semester):
-        return
+        return False
     # does the user have at least one required course in this semester?
-    required_courses = Course.objects.filter(participants=request.user, semester=semester, is_required_for_reward=True)
+    required_courses = Course.objects.filter(participants=user, semester=semester, is_required_for_reward=True)
     if not required_courses.exists():
-        return
+        return False
     # does the user not participate in any more required courses in this semester?
-    if required_courses.exclude(voters=request.user).exists():
-        return
+    if required_courses.exclude(voters=user).exists():
+        return False
     # did the user not already get reward points for this semester?
-    if RewardPointGranting.objects.filter(user_profile=request.user, semester=semester).exists():
-        return
+    if RewardPointGranting.objects.filter(user_profile=user, semester=semester).exists():
+        return False
     # grant reward points
-    RewardPointGranting.objects.create(user_profile=request.user, semester=semester, value=settings.REWARD_POINTS_PER_SEMESTER)
-    messages.success(request, _("You just have earned reward points for this semester because you evaluated all your courses. Thank you very much!"))
+    RewardPointGranting.objects.create(user_profile=user, semester=semester, value=settings.REWARD_POINTS_PER_SEMESTER)
+    return True
+
+# Signal handlers
+
+@receiver(Course.course_evaluated)
+def grant_reward_points_after_evaluate(sender, **kwargs):
+    request = kwargs['request']
+    semester = kwargs['semester']
+
+    if grant_reward_points(request.user, semester):
+        messages.success(request, _("You just have earned reward points for this semester because you evaluated all your courses. Thank you very much!"))
+
+@receiver(models.signals.m2m_changed, sender=Course.participants.through)
+def grant_reward_points_after_delete(instance, action, reverse, pk_set, **kwargs):
+    # if users do not need to evaluate a course anymore, they may have earned reward points
+    if action == "post_remove":
+        if reverse:
+            for semester in Semester.objects.filter(course__pk__in=pk_set):
+                grant_reward_points(instance, semester)
+        else:
+            for user in UserProfile.objects.filter(pk__in=pk_set):
+                grant_reward_points(user, instance.semester)
```

Comment:
i needed a while to understand this and would suggest the following to speed that up for others:

- to add comments behind the `if` and `else` that say `course got removed` and `participant got removed` respectively
- to add variables `user = instance` and `course = instance` in the two branches

i admit that after checking the documentation of `m2m_changed`, it's pretty clear what happens, but still, since we're not using this mechanic anywhere else, maybe additional help for the reader is still warranted. @He3lixxx opinions?

Reply:
to me it was clear what happens but with the suggested changes it should be obvious even without having the detailed model structure in mind, so that would be nice.

Reply:
Agreed, I found it hard to express myself with the different sides altered. Same for the test names.

Reply:
For the sake of making this complete, altough this answer is way too late: I agree with karyon.